File: src/main/java/java/math/cat/BigSet.java
package java.math.cat;

import java.util.Collection;
import java.util.Iterator;
import java.util.Set;

/**
 * An implementation of Set that cannot enumerate its elements.
 * AC is also optional here.
 * @param <T> element type
 *
 * @author Vlad Patryshev
 * All source code is stored on <a href="http://code.google.com/p/categories/">http://code.google.com/p/categories/</a>
 */
public abstract class BigSet<T> implements Set<T> {
  @Override
  public int size() {
    return Integer.MAX_VALUE;
  }

  @Override
  public boolean isEmpty() {
    throw new UnsupportedOperationException("Emptiness unknown");
  }

  @Override
  public Iterator<T> iterator() {
    throw new UnsupportedOperationException("This set is not enumerable");
  }

  @Override
  public Object[] toArray() {
    throw new UnsupportedOperationException("This set is not enumerable");
  }

  @Override
  public <E> E[] toArray(E[] a) {
    throw new UnsupportedOperationException("This set is not enumerable");
  }

  @Override
  public boolean add(T setMorphism) {
    throw new UnsupportedOperationException("This set is immutable");
  }

  @Override
  public boolean remove(Object o) {
    throw new UnsupportedOperationException("This set is immutable");
  }

  @Override
  public boolean containsAll(Collection<?> c) {
    for (Object a : c) {
      if (!contains(a)) {
        return false;
      }
    }
    return true;
  }

  @Override
  public boolean addAll(Collection<? extends T> c) {
    throw new UnsupportedOperationException("This set is immutable");
  }

  @Override
  public boolean retainAll(Collection<?> c) {
    throw new UnsupportedOperationException("This set is immutable");
  }

  @Override
  public boolean removeAll(Collection<?> c) {
    throw new UnsupportedOperationException("This set is immutable");
  }

  @Override
  public void clear() {
    throw new UnsupportedOperationException("This set is immutable");
  }

  /**
   * A big set of all finite sets in Java. This set is infinite, of course.
   */
  @SuppressWarnings("unchecked")
  public final static BigSet<Set> FINITE_SETS = new BigSet<Set>() {

    @Override
    public boolean contains(Object o) {
      return o instanceof Set && ((Set) o).size() < Integer.MAX_VALUE;
    }
  };
}

File: src/main/scala/math/cat/BigSet.scala
package math.cat

/**
  * An implementation of a large Set, can be non-enumerable.
  * AC is also optional here.
  *
  * @tparam T element type
  */
abstract class BigSet[T] extends Set[T] {
  thisOne =>
  override def size: Int = Sets.InfiniteSize

  def whoami: String = "A BIG SET"

  override def toString: String = whoami

  // TODO: figure out whether to kick this out.
  //  override def containsAll(c: Seq[T]): Boolean = c.forall(contains)
  //
  //  override def addAll(c: Seq[_ <: T]): Boolean = notMutable
  //
  //  override def retainAll(c: util.Collection[_]): Boolean = notMutable
  //
  //  override def removeAll(c: util.Collection[_]): Boolean = notMutable
  //
  //  override def clear(): Unit = notMutable

  def map[U](f: Functions.Bijection[T, U]): BigSet[U] = BigSet((u: U) => this contains (f unapply u))

  override def filter(p: T => Boolean): BigSet[T] = BigSet((t: T) => p(t) && (this contains t))

  override def hashCode: Int = System.identityHashCode(this)

  /**
    * This is a pretty bad equality for sets, but what can we do?
    * @param other anything, according to Java rules
    * @return true if it is the same set
    */
  override def equals(other: Any): Boolean = other match {
    case ref: AnyRef => this eq ref
    case notaref => false
  }
}

object BigSet {
  def apply[T](set: Set[T]): BigSet[T] = new BigSet[T] with EnumerableSet[T] {
    override def contains(t: T) = set(t)

    override def iterator: Iterator[T] = set.iterator
  }

  def apply[T](p: T => Boolean = (_: T) => true): BigSet[T] =
    new BigSet[T] with NonEnumerableSet[T] {
      override def contains(t: T) = p(t)
    }

}
